packages/@react-spectrum/table/src/DragPreview.tsx

```diff
@@ -0,0 +1,132 @@
+/*
+ * Copyright 2021 Adobe. All rights reserved.
+ * This file is licensed to you under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License. You may obtain a copy
+ * of the License at http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under
+ * the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
+ * OF ANY KIND, either express or implied. See the License for the specific language
+ * governing permissions and limitations under the License.
+ */
+import {classNames} from '@react-spectrum/utils';
+import {Direction} from '@react-types/shared';
+import {GridNode} from '@react-types/grid';
+import React from 'react';
+import {SpectrumTableProps} from '@react-types/table';
+import styles from '@adobe/spectrum-css-temp/components/table/vars.css';
+import stylesOverrides from './table.css';
+import type {TableLayout} from '@react-stately/layout';
+
+interface DragPreviewProps<T> {
+  item: GridNode<any>,
+  itemCount: number,
+  density: SpectrumTableProps<T>['density'],
+  layout: TableLayout<T>,
+  direction: Direction
+}
+
+interface TableCellPreviewProps {
+  cell: GridNode<any>,
+  layout: TableLayout<any>,
+  selectionCellWidth: number,
+  dragButtonCellWidth: number,
+  direction: Direction
+}
+
+export function TableCellPreview(props: TableCellPreviewProps) {
+  let {cell, layout, selectionCellWidth, dragButtonCellWidth, direction} = props;
+  let {align} = cell.column.props;
+  let {x: left, height, width} = layout.getLayoutInfo(cell.key).rect;
```

Comment:
if some columns are yet to be rendered and you're having trouble getting the width, you should have all column widths available through the layout state
`state.getColumnWidth(cell.column.key)`

height you may have to find a cell that has been rendered and grab that height and apply it to all of them

you could probably ignore left if you laid out the cells in a Flex where every item has a width and the flex prop is set to `flex: 0 0 auto`. this should address your RTL support as well, remove the need to determine badge width, and make it easier to vertically center the badge.

Reply:
Updated to use flex